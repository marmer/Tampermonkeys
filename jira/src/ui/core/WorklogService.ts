export interface Worklog {
    author: {
        displayName: string
    }
    timeSpendInMinutes: number
}

export default class WorklogService {
    getSummedWorklogsByUser(): Worklog[] {
        return [{author: {displayName: "me"}, timeSpendInMinutes: 42},
            {author: {displayName: "myself"}, timeSpendInMinutes: 546},
            {author: {displayName: "and I"}, timeSpendInMinutes: 1337}]
    }
}