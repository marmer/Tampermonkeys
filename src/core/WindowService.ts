export default class WindowService {
    public static getWindowLocationPathname(): string {
        return window.location.pathname;
    }
}