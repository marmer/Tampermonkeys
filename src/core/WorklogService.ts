import groupBy from "./groupBy";
import IssueSiteInfos from "./IssueSiteInfos";
import JiraTimeService from "./JiraTimeService";
import UserService from "./UserService";

export interface WorklogSumByUser {
    author: {
        displayName: string,
    };
    timeSpentInMinutes: number;
}

export interface Worklog {
    author: {
        displayName: string,
        name: string;
    };
    timeSpentInMinutes: number;
    started: string;
    id: string;
    comment: string;
    issueId: string;
}

export default class WorklogService {
    public static async createWorklog(worklog: {
        timeSpentInMinutes: number;
        started: string;
        comment: string;
        issueKey: string;
    }): Promise<void> {
        const {comment, issueKey, started, timeSpentInMinutes} = worklog;
        const timeSpent = JiraTimeService.minutesToJiraFormat(timeSpentInMinutes);

        const response = await fetch(IssueSiteInfos.getWorklogUrlForIssueKey(issueKey), {
            method: "POST",
            body: JSON.stringify({
                comment,
                timeSpent,
                started,
            }),
        }).catch(() => {
            throw new Error("Communication error: Worklog creation failed");
        });

        if (response.status !== 201) {
            throw new Error("Unexpected status code. Creation of worklog probably not sucecssful");
        }
    }

    public static async deleteWorklog(worklog: Worklog): Promise<void> {
        const response = await fetch(IssueSiteInfos.getWorklogModifyUrlByWorklog(worklog), {method: "DELETE"})
            .catch(() => {
                throw new Error("Communication error: Worklog creation failed");
            });

        if (response.status !== 204) {
            throw new Error("Unexpected status code. Deletion of worklog probably not sucecssful");
        }
    }

    public static async updateWorklog(worklog: Worklog): Promise<void> {
        // TODO: marmer 08.10.2019 implement
        throw new Error("Not implemented yet");
    }
    public static async getSummedWorklogsByUser(): Promise<WorklogSumByUser[]> {
        const issueKey = IssueSiteInfos.getCurrentIssueKey();
        return this.getWorklogByIssueKey(issueKey).then(this.sumUp);
    }

    public static async getWorklogsForCurrentIssueAndUser(): Promise<Worklog[]> {
        const worklogByIssueKey = await WorklogService.getWorklogByIssueKey(IssueSiteInfos.getCurrentIssueKey());
        const currentUserName = await UserService.getCurrentUserName();
        return worklogByIssueKey.filter(worklog => worklog.author.name === currentUserName);
    }

    private static async getWorklogByIssueKey(issueKey: string): Promise<Worklog[]> {
        return fetch(IssueSiteInfos.getWorklogUrlForIssueKey(issueKey), {method: "GET"})
            .then((response) => {
                if (response.status !== 200) {
                    throw new Error("Unexpected response status: " + response.status);
                }
                return response.json();
            })
            .then((responseJson) => responseJson.worklogs.map(WorklogService.toWorklog));
    }

    private static toWorklog(responseWorklog: any): Worklog {
        const {timeSpentSeconds, comment, author, id, started, issueId} = responseWorklog;
        const timeSpentInMinutes = Math.floor(timeSpentSeconds / 60);
        const {displayName, name} = author;

        return {
            author: {
                displayName,
                name,
            },
            timeSpentInMinutes,
            comment,
            id,
            started,
            issueId,
        };
    }

    private static sumUp(worklog: Worklog[]): WorklogSumByUser[] {
        const worklogSums: WorklogSumByUser[] = worklog.map(wl => ({
            author: {
                displayName: wl.author.displayName,
            },
            timeSpentInMinutes: wl.timeSpentInMinutes,
        }));

        return groupBy(worklogSums, (w) => w.author.displayName)
            .map((tupel) => tupel.values.reduce((previousValue, currentValue) => {
                const newValue = {...previousValue};
                newValue.timeSpentInMinutes += currentValue.timeSpentInMinutes;
                return newValue;
            }))
            .sort((a, b) => a.author.displayName < b.author.displayName ? -1 : 1);
    }
}
