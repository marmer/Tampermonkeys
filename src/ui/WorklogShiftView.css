.worklogShiftTarget {
    text-align: center;
    align-self: baseline !important;
    display: flex;
    flex-direction: column;
    margin-top: 0.2em;
}

.worklogShiftTarget > * {
    margin-top: 0.5em;
}
