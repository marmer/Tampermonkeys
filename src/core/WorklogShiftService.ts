import JiraTimeService from "./JiraTimeService";
import WorklogService, {Worklog} from "./WorklogService";

export default class WorklogShiftService {
    public static async shiftWorklog(worklog: Worklog, timeToShiftJiraString: string, targetIssueKey: string): Promise<void> {
        const timeToShiftInMinutes = JiraTimeService.jiraFormatToMinutes(timeToShiftJiraString);
        if (timeToShiftInMinutes === 0) {
            return;
        }
        if (timeToShiftInMinutes > worklog.timeSpentInMinutes) {
            throw new Error("It's not possible to shift more time than exist on a worklog");
        }

        const {
            comment,
            started,
            timeSpentInMinutes,
        } = worklog;

        await WorklogService.createWorklog({
            comment,
            issueKey: targetIssueKey,
            started,
            timeSpentInMinutes,
        }).catch(() => {
            throw new Error("Error while creating the new worklog. At least the source worklog has not been changed yet.");
        });

        try {
            if (timeSpentInMinutes === timeToShiftInMinutes) {
                await WorklogService.deleteWorklog(worklog);
            } else {
                await WorklogService.updateWorklog({
                    ...worklog,
                    timeSpentInMinutes: timeSpentInMinutes - timeToShiftInMinutes,
                });
            }
        } catch (reason) {
            throw new Error("Error while updating the worklog to change (source).");
        }
    }
}
