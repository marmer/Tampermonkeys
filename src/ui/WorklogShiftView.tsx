import moment = require("moment");
import React, {Component, ReactNode} from "react";
import JiraTimeService from "../core/JiraTimeService";
import WorklogService, {Worklog} from "../core/WorklogService";
import EstimationView from "./EstimationView";

// TODO: marmer 27.09.2019 care!
// tslint:disable-next-line:no-empty-interface
interface WorklogShiftViewState {
    worklogs: Worklog[];
}

export default class WorklogShiftView extends Component<{}, WorklogShiftViewState> {

    constructor(props: Readonly<{}>) {
        super(props);
        this.state = {
            worklogs: [],
        };
    }

    public componentDidMount(): void {
        WorklogService.getWorklogsForCurrentIssueAndUser()
            .then(worklogs => this.setState({worklogs}));
    }

    public render(): React.ReactElement {
        return <>
            {/*// TODO: marmer 30.09.2019 This is just a mockup with inline styles*/}
            {/*// TODO: marmer 30.09.2019 Don't use the layout of a different View in this way*/}
            This is just work in Progress
            <div className="estimationShiftContainer">
                <table>
                    <thead>
                    <tr>
                        <th>Author</th>
                        <th>Comment</th>
                        <th>Start</th>
                        <th>Time Spent</th>
                        <th>Time to move</th>
                        <th>Move</th>
                    </tr>
                    </thead>
                    <tbody>
                    {
                        this.state.worklogs.map(this.toRow)
                    }
                    </tbody>
                </table>
                <label style={{textAlign: "center", alignSelf: "baseline"}}>
                    Target
                    {/*// TODO: marmer 30.09.2019 Don't use the layout of a different View in this way*/}
                    <EstimationView className="estimationShiftCardContainer" readonly={true}
                                    estimation={{
                                        issueKey: "issue-1234",
                                        issueSummary: "something todo",
                                        originalEstimate: "1w 2d",
                                        remainingEstimate: "4h 15m",
                                    }}/></label>
            </div>
        </>;
    }

    private toRow(worklog: Worklog): ReactNode {

        return <tr key={worklog.id}>
            <td><input style={{textAlign: "center"}} type="text" disabled={true}
                       value={worklog.author.displayName}/></td>
            <td><input style={{textAlign: "center"}} type="text" disabled={true}
                       value={worklog.comment}/></td>
            <td><input style={{textAlign: "center"}} type="text" disabled={true}
                       value={moment(worklog.started).format("YYYY-MM-DD HH:mm:ss")}/></td>
            <td><input style={{textAlign: "center"}} type="text" disabled={true}
                       value={JiraTimeService.minutesToJiraFormat(worklog.timeSpentInMinutes)}/></td>
            <td><input style={{textAlign: "center"}} type="text" placeholder="5h 9m"
                       value={JiraTimeService.minutesToJiraFormat(worklog.timeSpentInMinutes)}/>
            </td>
            <td>
                <button title="move">{">"}</button>
            </td>
        </tr>;
    }
}
