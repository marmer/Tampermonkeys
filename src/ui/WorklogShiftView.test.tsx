import * as reactTest from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import React from "react";
import WindowService from "../core/WindowService";
import WorklogService, {Worklog} from "../core/WorklogService";
import WorklogShiftService from "../core/WorklogShiftService";
import WorklogShiftView from "./WorklogShiftView";

describe("WorklogShiftView", () => {

    const worklogBase: Worklog = {
        issueId: "42",
        author: {
            name: "real.me",
            displayName: "Me",
        },
        timeSpentInMinutes: 1337,
        started: "2000-01-01T00:00:00.000+0000",
        id: "43",
        comment: "did something",
    };

    it("should not show a worklog while it's loading", async () => {
        WorklogService.getWorklogsForCurrentIssueAndUser = jest.fn().mockResolvedValue([{...worklogBase}] as Worklog[]);

        const underTest = reactTest.render(<WorklogShiftView/>);
        expect(underTest).toMatchSnapshot();
    });

    it("should show the worklogs for the current user and issue when loading is done", async () => {
        WorklogService.getWorklogsForCurrentIssueAndUser = jest.fn().mockResolvedValue([
            {
                ...worklogBase,
                author: {...worklogBase.author, displayName: "Me"},
                comment: "first step",
                started: "2019-10-01T01:02:03.000+0200",
                timeSpentInMinutes: 1,
                id: "1111",
                issueId: "11",
            },
            {
                ...worklogBase,
                author: {...worklogBase.author, displayName: "Me Too"},
                comment: "second step",
                started: "2020-01-10T02:03:04.000+0200",
                timeSpentInMinutes: 2,
                id: "2222",
                issueId: "22",
            }] as Worklog[]);

        const underTest = reactTest.render(<WorklogShiftView/>);
        await reactTest.waitForElement(() => underTest.getByText("first step"));
        expect(underTest).toMatchSnapshot();
    });

    it("should describe the absence of own worklogs", async () => {
        WorklogService.getWorklogsForCurrentIssueAndUser = jest.fn().mockResolvedValue([] as Worklog[]);

        const underTest = reactTest.render(<WorklogShiftView/>);
        await reactTest.waitForElement(() => underTest.getByText("No work logged here for you yet"));
        expect(underTest).toMatchSnapshot();
    });

    it("should tell the user when an error has occurred", async () => {
        WorklogService.getWorklogsForCurrentIssueAndUser = jest.fn().mockRejectedValue(new Error("dafuq"));

        const underTest = reactTest.render(<WorklogShiftView/>);
        await reactTest.waitForElement(() => underTest.getByText("Error while loading worklogs for issue: dafuq"));
        expect(underTest).toMatchSnapshot();
    });

    it("should reload the current site after some worklog has been shifted", async () => {
        const sourceWorklog = {...worklogBase};
        WorklogService.getWorklogsForCurrentIssueAndUser = jest.fn().mockResolvedValue([sourceWorklog] as Worklog[]);
        WorklogShiftService.shiftFromWorklog = jest.fn().mockResolvedValue(undefined);
        WindowService.reloadPage = jest.fn();

        const underTest = reactTest.render(<WorklogShiftView/>);
        const targetIssueInput = await reactTest.waitForElement(() => underTest.getByTitle("Target Issue"));

        userEvent.type(targetIssueInput, "TARGET-123");
        userEvent.type(underTest.getByTestId("ShiftInput" + worklogBase.id), "5m");
        const shiftButton = underTest.getByTestId("ShiftButton" + worklogBase.id);
        userEvent.click(shiftButton);

        await reactTest.wait(() => expect(WorklogShiftService.shiftFromWorklog).toBeCalledWith(sourceWorklog, "5m", "TARGET-123"));
        expect(WindowService.reloadPage).toBeCalled();
    });

    // TODO: marmer 08.10.2019 default values
    // TODO: marmer 07.10.2019 errorhandling while shifting
    // TODO: marmer 07.10.2019 Handling of missing worklog parts (author, Comment, Start, ...)
    // TODO: marmer 07.10.2019 Keep destination issue in session storage in case the user wants to shift more than one worklog
    // TODO: marmer 07.10.2019 Shifting should only be possible if the target issue exists
    // TODO: marmer 07.10.2019 show some target issue details
    // TODO: marmer 07.10.2019 handle errors while loading the target issue
});
