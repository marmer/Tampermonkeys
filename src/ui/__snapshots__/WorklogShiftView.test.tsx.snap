// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorklogShiftView should show the estimations for the current user and issue when loading is done 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      This is just work in Progress
      <div
        class="estimationShiftContainer"
      >
        <table>
          <thead>
            <tr>
              <th>
                Author
              </th>
              <th>
                Comment
              </th>
              <th>
                Start
              </th>
              <th>
                Time Spent
              </th>
              <th>
                Time to move
              </th>
              <th>
                Move
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                <input
                  disabled=""
                  style="text-align: center;"
                  type="text"
                  value="Me"
                />
              </td>
              <td>
                <input
                  disabled=""
                  style="text-align: center;"
                  type="text"
                  value="first step"
                />
              </td>
              <td>
                <input
                  disabled=""
                  style="text-align: center;"
                  type="text"
                  value="2019-10-01 01:02:03"
                />
              </td>
              <td>
                <input
                  disabled=""
                  style="text-align: center;"
                  type="text"
                  value="1m"
                />
              </td>
              <td>
                <input
                  placeholder="5h 9m"
                  style="text-align: center;"
                  type="text"
                  value="1m"
                />
              </td>
              <td>
                <button
                  title="move"
                >
                  &gt;
                </button>
              </td>
            </tr>
            <tr>
              <td>
                <input
                  disabled=""
                  style="text-align: center;"
                  type="text"
                  value="Me Too"
                />
              </td>
              <td>
                <input
                  disabled=""
                  style="text-align: center;"
                  type="text"
                  value="second step"
                />
              </td>
              <td>
                <input
                  disabled=""
                  style="text-align: center;"
                  type="text"
                  value="2020-01-10 01:03:04"
                />
              </td>
              <td>
                <input
                  disabled=""
                  style="text-align: center;"
                  type="text"
                  value="2m"
                />
              </td>
              <td>
                <input
                  placeholder="5h 9m"
                  style="text-align: center;"
                  type="text"
                  value="2m"
                />
              </td>
              <td>
                <button
                  title="move"
                >
                  &gt;
                </button>
              </td>
            </tr>
          </tbody>
        </table>
        <label
          style="text-align: center; align-self: baseline;"
        >
          Target
          <div
            class="estimationShiftCardContainer"
            title="issue-1234: something todo"
          >
            <label>
              Issue 
              <input
                disabled=""
                type="text"
                value="issue-1234"
              />
            </label>
            <label>
              Original Estimate 
              <input
                disabled=""
                type="text"
                value="1w 2d"
              />
            </label>
            <label>
              Remaining Estimate 
              <input
                disabled=""
                type="text"
                value="4h 15m"
              />
            </label>
          </div>
        </label>
      </div>
    </div>
  </body>,
  "container": <div>
    This is just work in Progress
    <div
      class="estimationShiftContainer"
    >
      <table>
        <thead>
          <tr>
            <th>
              Author
            </th>
            <th>
              Comment
            </th>
            <th>
              Start
            </th>
            <th>
              Time Spent
            </th>
            <th>
              Time to move
            </th>
            <th>
              Move
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              <input
                disabled=""
                style="text-align: center;"
                type="text"
                value="Me"
              />
            </td>
            <td>
              <input
                disabled=""
                style="text-align: center;"
                type="text"
                value="first step"
              />
            </td>
            <td>
              <input
                disabled=""
                style="text-align: center;"
                type="text"
                value="2019-10-01 01:02:03"
              />
            </td>
            <td>
              <input
                disabled=""
                style="text-align: center;"
                type="text"
                value="1m"
              />
            </td>
            <td>
              <input
                placeholder="5h 9m"
                style="text-align: center;"
                type="text"
                value="1m"
              />
            </td>
            <td>
              <button
                title="move"
              >
                &gt;
              </button>
            </td>
          </tr>
          <tr>
            <td>
              <input
                disabled=""
                style="text-align: center;"
                type="text"
                value="Me Too"
              />
            </td>
            <td>
              <input
                disabled=""
                style="text-align: center;"
                type="text"
                value="second step"
              />
            </td>
            <td>
              <input
                disabled=""
                style="text-align: center;"
                type="text"
                value="2020-01-10 01:03:04"
              />
            </td>
            <td>
              <input
                disabled=""
                style="text-align: center;"
                type="text"
                value="2m"
              />
            </td>
            <td>
              <input
                placeholder="5h 9m"
                style="text-align: center;"
                type="text"
                value="2m"
              />
            </td>
            <td>
              <button
                title="move"
              >
                &gt;
              </button>
            </td>
          </tr>
        </tbody>
      </table>
      <label
        style="text-align: center; align-self: baseline;"
      >
        Target
        <div
          class="estimationShiftCardContainer"
          title="issue-1234: something todo"
        >
          <label>
            Issue 
            <input
              disabled=""
              type="text"
              value="issue-1234"
            />
          </label>
          <label>
            Original Estimate 
            <input
              disabled=""
              type="text"
              value="1w 2d"
            />
          </label>
          <label>
            Remaining Estimate 
            <input
              disabled=""
              type="text"
              value="4h 15m"
            />
          </label>
        </div>
      </label>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
